### PR TITLE
feat: optimize loser tree, disable peek top2

### PR DESCRIPTION
**Pull request by forsaken628**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary
Simply disable peek top2 for loser tree, but it seems to work? NEED more test case.

link #15925

[pprof.samples.cpu.heap.pb.gz](https://github.com/user-attachments/files/16100670/pprof.samples.cpu.heap.pb.gz)
heap

[pprof.samples.cpu.loser.pb.gz](https://github.com/user-attachments/files/16100671/pprof.samples.cpu.loser.pb.gz)
loser

### data set
tpc-h 10G
### sql
select * from orders order by o_orderdate desc, o_totalprice desc ignore_result;

## Tests

- [ ] Unit Test
- [ ] Logic Test
- [x] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [x] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/15979)

### CODE REVIEW

**Comment by forsaken628** [comment]
commit: 100434a202da2aa7b517acd37c0c2153c22857f1

setting:
set sort_spilling_memory_ratio = 0;
set sort_spilling_bytes_threshold_per_proc = 0;

sql: select * from orders order by o_orderdate desc, o_totalprice desc ignore_result;

|| heap      | loser tree |remark|
|-| ----------- | ----------- |-|
|Total CPU| 50.13s      | 42.93s ||
|Running Time|31.116s|29.214s|
|MultiSortMerge| 6.38s   | 6.29s ||
|MergeSort.on_finish| 36.22s   | 31.63s        ||
|evaluate_cursor| 12.66s   | 8.11s |for all two merges|
|BinaryColumn::row| 5.65s   | 2.52s | for all loading rows before comparing|

[pprof.cpu.heap.pb.gz](https://github.com/user-attachments/files/16246520/pprof.cpu.heap.pb.gz)
[pprof.cpu.loser_tree.pb.gz](https://github.com/user-attachments/files/16246521/pprof.cpu.loser_tree.pb.gz)

**Comment by forsaken628** [comment]
I found an issue with sorting stability that seems undefined
input:
block(0) (2, ...)
block(1) (1,2, ...)
output:
(1,0) (0,1) (1,1) ... // min input win
(1,0) (1,1) (0,1) ... // same input win

This is min input win
https://github.com/datafuselabs/databend/blob/3d47fcfd03dce6229c39f59f301bed960486e7cc/src/query/pipeline/transforms/src/processors/transforms/sort/cursor.rs#L117-L125

This is same input win
https://github.com/datafuselabs/databend/blob/3d47fcfd03dce6229c39f59f301bed960486e7cc/src/query/pipeline/transforms/src/processors/transforms/sort/merger.rs#L173

**Comment by forsaken628** [comment]
Clean up redundant judgments because stable ordering is meaningless

**Comment by sundy-li** [comment]
>  because stable ordering is meaningless

It's correct in distributed dbms.

File: src/query/pipeline/transforms/src/processors/transforms/sort/cursor.rs [hunk 1]
```diff
@@ -78,23 +78,25 @@ impl<R: Rows> Cursor<R> {
             cursor: self,
         }
     }
+
+    fn same_index(&self, other: &Self) -> bool {
+        self.input_index == other.input_index && self.row_index == other.row_index
```
Comment (sundy-li) [inline src/query/pipeline/transforms/src/processors/transforms/sort/cursor.rs]: What's the case that two cursors will share `same_index` ?

File: src/query/pipeline/transforms/src/processors/transforms/sort/merger.rs [hunk 2]
```diff
@@ -165,73 +165,57 @@ where
...
+        let input_index = cursor.input_index;
+        let start = cursor.row_index;
+
         let max_rows = self.limit.unwrap_or(self.batch_rows).min(self.batch_rows);
-        let (whole_block, next_cursor) = if self.sorted_cursors.len() == 1 {
-            (true, None)
+        let count = if self.sorted_cursors.len() == 1 {
+            (cursor.num_rows() - start).min(max_rows - self.temp_sorted_num_rows)
+        } else if !A::SHOULD_PEEK_TOP2 {
+            debug_assert!(!cursor.is_finished());
```
Comment (sundy-li) [inline src/query/pipeline/transforms/src/processors/transforms/sort/merger.rs:175]: In this code block, we continue to find the smallest row with equal.

So I think both heap and loser tree can share this branch. It's nothin with `SHOULD_PEEK_TOP2`

Comment (sundy-li) [reply src/query/pipeline/transforms/src/processors/transforms/sort/merger.rs:175]: Where is the top2 cursor now?

Comment (sundy-li) [reply src/query/pipeline/transforms/src/processors/transforms/sort/merger.rs:175]: I think it's also right to advance he heap to find the equal element.

**Comment by forsaken628** [comment]
> > because stable ordering is meaningless
> 
> It's correct in distributed dbms.

The current Cursor order logic is strongly coupled with the TransformSortMergeLimit implementation. But Cursor looks generic, so it's easy to get confused.

File: src/query/pipeline/transforms/src/processors/transforms/sort/merger.rs (hunk 2 above)
Comment (forsaken628) [reply src/query/pipeline/transforms/src/processors/transforms/sort/merger.rs]: Consider the following example:
cursor_top_1:  [1,1,2,3 ...]
cursor_top_2:  [2,...]

for !SHOULD_PEEK_TOP2 output [1,1]
for SHOULD_PEEK_TOP2 output [1,1,2] 

The heap branch is not worse.

**Comment by forsaken628** [comment]
```
INSERT INTO t1 VALUES (1),(1),(1),(3),(3),(5),(5)

SELECT a, sum(a) OVER w FROM t1 WINDOW w AS (PARTITION BY a) ORDER BY count() OVER w DESC
expected:
1 3
1 3
1 3
3 6
3 6
5 10
5 10
----
actual:
1 3
1 3
1 3
5 10
5 10
3 6
3 6
```
The test failed, but this was expected due to unstable sorting.

**Comment by sundy-li** [comment]
> The test failed, but this was expected due to unstable sorting.

Window operators don't ensure the ordering of results.

Let's add an order by  in SQL to keep the result stable.

**Review by sundy-li — APPROVED** [review]

**Pull request merged**